Does this pull request fix issue #24 ?

jelber2 opened the pull request:
Does this pull request fix issue #24  ?

get cDNA
```bash
wget ftp://ftp.ensemblgenomes.org/pub/metazoa/release-46/fasta/caenorhabditis_elegans/cdna/Caenorhabditis_elegans.WBcel235.cdna.all.fa.gz
```

use seqtk to isolate first cDNA sequence 
```bash
seqtk seq -l0 Caenorhabditis_elegans.WBcel235.cdna.all.fa.gz |paste - - |head -n1 |tr '\t' '\n' > test.fa
```

rust setup
```bash
rustup show

Default host: x86_64-unknown-linux-gnu
rustup home:  /nfs/scistore16/itgrp/jelbers/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.60.0 (7737e0b5c 2022-04-04)
```

mudskipper installation
```bash
cd ~/git/
git clone https://github.com/jelber2/mudskipper
cd mudskipper
git show

commit 993320bdc4528615311c2250d68c3d7a91c2efc4 (HEAD -> main, origin/main, origin/HEAD)

cargo build --release
```

get GTF file and genome
```bash
wget ftp://ftp.ensemblgenomes.org/pub/metazoa/release-46/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.46.gtf.gz
gunzip Caenorhabditis_elegans.WBcel235.46.gtf.gz
wget ftp://ftp.ensemblgenomes.org/pub/metazoa/release-46/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna.toplevel.fa.gz
gunzip Caenorhabditis_elegans.WBcel235.dna.toplevel.fa.gz
```

STAR index genome
```bash
module load seqtk/20210125
# make sequences upper case (remove soft-masking)
seqtk seq -CU Caenorhabditis_elegans.WBcel235.dna.toplevel.fa > Caenorhabditis_elegans.WBcel235.dna.toplevel_uppercase.fa

module load STAR/2.7.9a
mkdir -p Caenorhabditis_elegans.WBcel235.dna.toplevel_Ensembl46_STAR-2.7.9a_sjdbOverhang18
STAR \
--genomeSAindexNbases 12 \
--limitGenomeGenerateRAM 100000000000 \
--runThreadN 34 \
--sjdbOverhang 18 \
--runMode genomeGenerate \
--genomeDir Caenorhabditis_elegans.WBcel235.dna.toplevel_Ensembl46_STAR-2.7.9a_sjdbOverhang18 \
--genomeFastaFiles Caenorhabditis_elegans.WBcel235.dna.toplevel_uppercase.fa \
--sjdbGTFfile Caenorhabditis_elegans.WBcel235.46.gtf
```

simulate one 150-bp read pair on the transcript Y110A7A.10.1 with BBTools/BBMap's randomreads.sh version 38.82 (any recent version should be fine too)
```bash
module load bbtools/38.82
randomreads.sh ref=test.fa out1=test_R1.fq.gz out2=test_R2.fq.gz \
reads=1 simplenames=t illuminanames=t paired=t addslash=t seed=1
```
test_R1.fq.gz 
```bash
@0/1
ACAATATTTGATGTGTGGAACTTCTAAGGAAAATGCTGCGAATGCTCTGATTCATGCTCGCTATCGGATCGCTCAACTTGACAAGGCTGTTGGATTGAAAAGAGAGCCACAAGATGGTATTTTCTTGATTCGAGGATCCTCTTCTCATGG
+
6=5559=;9=A<>A?=@A?CD@A=D@B@=@?DDDCB?ABD<D>CBA?BDB;@?AABD>@B@:C?A>=9:???@:D?8B?C>@:>@C;@=<??=@AA>?7?ABA6?@A?E@?;>B>B;7<<=7A9C:><=:<@;9C;:<;<5555;67:65
```

test_R2.fq.gz
```bash
@0/2
GTTTTGATAGCATCCGCGTGGGTTTCAAGACTGAAGTGAGAGTAATAGCGAACGAAGTCACTTATAGTTGTCAGGTAGACATTGCTGTGCTCGAATCCCCATCCTTCTTCGTTTTGTTCGATTAGACAATGTGAGACACGTTCTCCGTGA
+
44788<4<<;9>>??B@AB@<<>6?>?=BAAC5=;<@><;5@?C?BA8B;>?>>>=;A<C@?;?>BAA??C>CA?B96@6:@@=>:@<B<>@AB@<:><<>C;=C6@?=@9@7?==6<?<:9@A:?>7:=?A;8=A995<444;;44446
```

STAR mapping
```bash
module load STAR/2.7.9a
STAR  --runThreadN 32 \
--genomeDir auxData/Caenorhabditis_elegans.WBcel235.dna.toplevel_Ensembl46_STAR-2.7.9a_sjdbOverhang18 \
--readFilesCommand zcat \
--outSAMtype BAM SortedByCoordinate \
--outFileNamePrefix mapped/ref/RNAseq.MUT1.rep4 \
--outTmpDir TMP/ref_RNAseq.MUT1.rep4/TEMP/ \
--outReadsUnmapped Fastx \
--readFilesIn test_R1.fastq.gz test_R2.fastq.gz
```

mudskipper index GTF
```bash
export PATH="/nfs/scistore16/itgrp/jelbers/git/mudskipper/target/release:$PATH"
mudskipper index --gtf auxData/Caenorhabditis_elegans.WBcel235.46.gtf --dir-index auxData/mudskipper

[2022-05-12T06:02:52Z INFO  mudskipper] Mudskipper started...
[2022-05-12T06:02:52Z INFO  mudskipper::annotation] reading the gtf file and building the tree.
[2022-05-12T06:02:59Z INFO  mudskipper::annotation] building the tree
[2022-05-12T06:02:59Z INFO  mudskipper::annotation] Time to build the tree: 7.234329464s
[2022-05-12T06:02:59Z INFO  mudskipper::annotation] saving the GTF index
[2022-05-12T06:03:03Z INFO  mudskipper::annotation] Done with saving the GTF index
[2022-05-12T06:03:03Z INFO  mudskipper] Mudskipper finished.
```

run mudskipper bulk
```
mudskipper bulk --threads 32 --alignment Aligned.sortedByCoord.out.bam --index auxData/mudskipper --out alignments.bam
[2022-05-17T08:21:36Z INFO  mudskipper] Mudskipper started...
[2022-05-17T08:21:36Z INFO  mudskipper::annotation] Loading parsed GTF...
[2022-05-17T08:21:36Z INFO  mudskipper::bam] Number of reference sequences: 61451
[2022-05-17T08:21:36Z INFO  mudskipper::bam] thread count: 32 in total
[2022-05-17T08:21:36Z INFO  mudskipper::bam] thread count: 16 for reading
[2022-05-17T08:21:36Z INFO  mudskipper::bam] thread count: 16 for writing
[2022-05-17T08:21:37Z INFO  mudskipper] Mudskipper finished.
```
Aligned.sortedByCoord.out.bam (no header)
```bash
0	99	I	5108012	255	5M47N145M	=	5108277	596	GTGAAATTTGCCGCCCAAGTGAAACTTCTCCGAATCGTTGTTAAAGATGGAAAATGCGGATTCAATACAGATTCATTAACTCACGATTCTGTTGTCAGAGTAATTGAATTTCATCGAAATATTTCTCTCAATATTTTCAATGATGCACTT	8;9=898:;@>B9DAD89AB;C?EDEH<?B8AE9AD>?ACBC8>:@98=CFB?A;A=?9>:D;@BAAE>>@8A<>@=<AE=G>:C@@8;;8B=>>D@E?88=;CA89::;?ABD>=@8:><8=C=E?@888?@A8D<8C@9=8?888888	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:99T50	jM:B:c,21	jI:B:i,5108017,5108063	MC:Z:47M181N103M
0	147	I	5108277	255	47M181N103M	=	5108012	-596	GGGGCCACTTCAGCAGCGTATGATTCTCAGTGCAAAAAATGATCATGATTGGAGAGAACGCTTGGAAATGGAAAATTTGAGAGCAGTACATTTGGCATTTGAGAGAGGAGCCAAGCTATATGATTCAGCTCATCAAGAAATGGAACGAGC	33338333347333383><37;3;398:8:79<785786=6<333?=77???93A;5A=B:79>48>A83;;3678<>@7A=:::=34337=??=;9BA<3=3;>83:=;>;5<9>9=?=;@9;;A=@;@=<B5:878:=8848:93533	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:5A144	jM:B:c,21	jI:B:i,5108324,5108504	MC:Z:5M47N145M
```

alignments.bam (no header)
```bash
0	99	Y110A7A.10.1	180	255	150M	=	398	368	GTGAAATTTGCCGCCCAAGTGAAACTTCTCCGAATCGTTGTTAAAGATGGAAAATGCGGATTCAATACAGATTCATTAACTCACGATTCTGTTGTCAGAGTAATTGAATTTCATCGAAATATTTCTCTCAATATTTTCAATGATGCACTT	8;9=898:;@>B9DAD89AB;C?EDEH<?B8AE9AD>?ACBC8>:@98=CFB?A;A=?9>:D;@BAAE>>@8A<>@=<AE=G>:C@@8;;8B=>>D@E?88=;CA89::;?ABD>=@8:><8=C=E?@888?@A8D<8C@9=8?888888	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:99T50	jM:B:c,21	jI:B:i,5108017,5108063	MC:Z:47M181N103M
0	147	Y110A7A.10.1	398	255	150M	=	180	-368	GGGGCCACTTCAGCAGCGTATGATTCTCAGTGCAAAAAATGATCATGATTGGAGAGAACGCTTGGAAATGGAAAATTTGAGAGCAGTACATTTGGCATTTGAGAGAGGAGCCAAGCTATATGATTCAGCTCATCAAGAAATGGAACGAGC	33338333347333383><37;3;398:8:79<785786=6<333?=77???93A;5A=B:79>48>A83;;3678<>@7A=:::=34337=??=;9BA<3=3;>83:=;>;5<9>9=?=;@9;;A=@;@=<B5:878:=8848:93533	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:5A144	jM:B:c,21	jI:B:i,5108324,5108504	MC:Z:5M47N145M
0	99	Y110A7A.10.2	170	255	150M	=	388	368	GTGAAATTTGCCGCCCAAGTGAAACTTCTCCGAATCGTTGTTAAAGATGGAAAATGCGGATTCAATACAGATTCATTAACTCACGATTCTGTTGTCAGAGTAATTGAATTTCATCGAAATATTTCTCTCAATATTTTCAATGATGCACTT	8;9=898:;@>B9DAD89AB;C?EDEH<?B8AE9AD>?ACBC8>:@98=CFB?A;A=?9>:D;@BAAE>>@8A<>@=<AE=G>:C@@8;;8B=>>D@E?88=;CA89::;?ABD>=@8:><8=C=E?@888?@A8D<8C@9=8?888888	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:99T50	jM:B:c,21	jI:B:i,5108017,5108063	MC:Z:47M181N103M
0	147	Y110A7A.10.2	388	255	150M	=	170	-368	GGGGCCACTTCAGCAGCGTATGATTCTCAGTGCAAAAAATGATCATGATTGGAGAGAACGCTTGGAAATGGAAAATTTGAGAGCAGTACATTTGGCATTTGAGAGAGGAGCCAAGCTATATGATTCAGCTCATCAAGAAATGGAACGAGC	33338333347333383><37;3;398:8:79<785786=6<333?=77???93A;5A=B:79>48>A83;;3678<>@7A=:::=34337=??=;9BA<3=3;>83:=;>;5<9>9=?=;@9;;A=@;@=<B5:878:=8848:93533	NH:i:1	HI:i:1	AS:i:298	nM:i:2	NM:i:1	MD:Z:5A144	jM:B:c,21	jI:B:i,5108324,5108504	MC:Z:5M47N145M
```

jelber2 committed:
Update query_bam_records.rs

jelber2 commented:
On second thought, perhaps my quick fix my only work with when 100% reads are paired and 100% mates are mapped?

rob-p commented:
Thanks for this test case and PR @jelber2! @gmarcais, can we take a look at this?

opplatek commented:
> On second thought, perhaps my quick fix may only work with when 100% reads are paired and 100% mates are mapped? I have not tested how it works when one mate is unmapped. **EDIT** if one mate is unmapped, the mapped mate is not transferred when using `mudskipper bulk`

I am pretty sure this is not just because of your fix but rather a combination with the rest of the `mudskipper bulk` code. Your solution works only if **both mates** are present in a **name sorted** SAM/BAM. If the SAM/BAM file is not name sorted `mudskipper bulk` treats all the reads as sort of pseudo-single-end reads - it duplicates each of the reads with the only difference being the `TLEN` sign (switches + for - and vice versa). 

For example (modified from tests/NC_007114.7_12427.sam):
1) Paired, mapped, name sorted = works fine (output not copied)
```
J00153:160:HMTNMBBXX:6:1104:18893:12427	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```
2) Paired, mapped, **not** name sorted (but coordinate sorted) = doesn't work:
```
J00153:160:HMTNMBBXX:6:1104:18893:12427	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```
results in:
```
J00153:160:HMTNMBBXX:6:1104:18893:12427	419	XM_021473592.1	1294	3	149M	=	1294	149	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	419	XM_021473592.1	1294	3	149M	=	1294	-149	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1294	149	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1294	-149	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	1408	150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	1408	-150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1408	150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1408	-150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```
3) One pair is **mapped** and **both mates** are **present**, the other pair is **missing a mate** completely (due to post-alignment filtering) but SAM/BAM is name sorted = the complete pair is correctly converted, the *one-mate-pair* read gets copied again with only the `TLEN` sign changed:
```
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
``` 
results in:
```
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1408	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1294	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	1408	150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	1408	-150	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```

jelber2 commented:
Dear Jan, thanks for catching this. I can see these trends in my 'real' data too. Using this pull request as is.

jelber2 committed:
Merge branch 'OceanGenomics:main' into main

jelber2 committed:
Update query_bam_records.rs

jelber2 committed:
Create NC_007114.7_12427_1.sam

jelber2 committed:
Create NC_007114.7_12427_2.sam

jelber2 committed:
Create NC_007114.7_12427_3.sam

jelber2 committed:
Update NC_007114.7_12427_3.sam

jelber2 commented:
Note that I added your test examples in `tests/`


1. Paired, mapped, name sorted
`target/release/mudskipper bulk --alignment tests/NC_007114.7_12427_1.sam --gtf tests/NC_007114.7_12427.gtf --out test1.bam`

```sh
cat <(grep -v "^@" tests/NC_007114.7_12427_1.sam) <(echo) <(samtools view test1.bam)

J00153:160:HMTNMBBXX:6:1104:18893:12427	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0

J00153:160:HMTNMBBXX:6:1104:18893:12427	419	XM_021473592.1	1294	3	149M	=	1408	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	1294	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1408	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1294	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```
2. Paired, mapped, cooridnate **has incorrect coordinates**
`target/release/mudskipper bulk --alignment tests/NC_007114.7_12427_2.sam --gtf tests/NC_007114.7_12427.gtf --out test2.bam`

```sh
cat <(grep -v "^@" tests/NC_007114.7_12427_2.sam) <(echo) <(samtools view test2.bam)

J00153:160:HMTNMBBXX:6:1104:18893:12427	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0

J00153:160:HMTNMBBXX:6:1104:18893:12427	419	XM_021473592.1	1294	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```

3. One pair is mapped and both mates are present, the other pair is missing a mate completely (due to post-alignment filtering) but SAM/BAM is name sorted = the complete pair is correctly converted, the one-mate-pair read **has incorrect coordinates**
`target/release/mudskipper bulk --alignment tests/NC_007114.7_12427_3.sam --gtf tests/NC_007114.7_12427.gtf --out test3.bam`

```sh
cat <(grep -v "^@" tests/NC_007114.7_12427_3.sam) <(echo) <(samtools view test3.bam)

J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0

J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1408	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1294	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	XM_021473592.1	1408	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```

jelber2 committed:
Update convert.rs

jelber2 committed:
Update query_bam_records.rs

jelber2 commented:
Ok, well, I think I fixed it for now, but one must still do sorting by read name prior to running mudskipper.

2. Paired, mapped, coordinate-sorted **must sort by read name, otherwise, pairs not next to each other are considered single-end reads**
`target/release/mudskipper bulk --alignment tests/NC_007114.7_12427_2.sam --gtf tests/NC_007114.7_12427.gtf --out test2.bam`

```sh
cat <(grep -v "^@" tests/NC_007114.7_12427_2.sam) <(echo) <(samtools view test2.bam)

J00153:160:HMTNMBBXX:6:1104:18893:12427	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0

J00153:160:HMTNMBBXX:6:1104:18893:12427	418	XM_021473592.1	1294	3	149M	*	0	0	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	418	XM_021473592.1	1294	3	149M	*	0	0	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	338	XM_021473592.1	1408	3	150M	*	0	0	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	338	XM_021473592.1	1408	3	150M	*	0	0	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```


3. One pair is mapped and both mates are present, the other pair is missing a mate completely (due to post-alignment filtering) but SAM/BAM is name sorted = the complete pair is correctly converted, the one-mate-pair read **will be considered a single-end read**
`target/release/mudskipper bulk --alignment tests/NC_007114.7_12427_3.sam --gtf tests/NC_007114.7_12427.gtf --out test3.bam`

```sh
cat <(grep -v "^@" tests/NC_007114.7_12427_3.sam) <(echo) <(samtools view test3.bam)

J00153:160:HMTNMBBXX:6:1104:18893:12428	419	NC_007114.7	24289610	3	149M	=	24289724	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	339	NC_007114.7	24289724	3	150M	=	24289610	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0

J00153:160:HMTNMBBXX:6:1104:18893:12428	419	XM_021473592.1	1294	3	149M	=	1408	264	GGCAGTGGTCTCTGTCCTGGATCATGTAGAGTGTCGTTGTCAACCTGCTCCTCGCCCCGCACATCGCCGAAAGTCATCCAGCCAGAAACTAGACGCACGTGATCGTTCAGAGAAAAGTCGACCCAAAGAGGATCTCCATCGCAGGGATG	AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJFFJJJFJJJJJJJJJFAFAFJJJJJJJJJA<A	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12428	339	XM_021473592.1	1408	3	150M	=	1294	-264	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
J00153:160:HMTNMBBXX:6:1104:18893:12427	338	XM_021473592.1	1408	3	150M	*	0	0	AAGTCGACCCAAAGAGGATCTCCATCGCAGGGATGAACTCAAACACAACCAGAGGCTAAACCTGGAGGACCTGCTCAGCCACAGCTGGCTGCCCCAAGACAGGTTCTCTGAATCACCAGACACTGTCCATTTTGGCCGCGATGCCTGGAC	FAFJJFAJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJJJJJJJJJJJJFJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA	NH:i:2	HI:i:2	AS:i:297	nM:i:0
```

jelber2 committed:
produce warning of missing mate

produce warning not error, like https://github.com/OceanGenomics/mudskipper/pull/29
